These changes were backported from PR #14378 that had originally been made by Cedric Champeau <cedric@gradle.com>
diff --git a/buildSrc/subprojects/dependency-modules/src/main/kotlin/gradlebuild/modules/extension/ExternalModulesExtension.kt b/buildSrc/subprojects/dependency-modules/src/main/kotlin/gradlebuild/modules/extension/ExternalModulesExtension.kt
index 7107bad26967c..1f90a64cd2931 100644
--- a/gradle/dependencies.gradle
+++ b/gradle/dependencies.gradle
@@ -33,6 +33,7 @@
 versions.commons_io = '2.5'
 versions.groovy = groovyVersion
 versions.bouncycastle = '1.58'
+versions.jettyVersion = "9.4.31.v20200723"
 versions.maven = '3.0.4'
 versions.httpclient = '4.4.1'
 
@@ -136,9 +136,14 @@
 libraries.logback_classic = 'ch.qos.logback:logback-classic:1.0.13'
 
 // Jetty
-libraries.servlet_api = "org.mortbay.jetty:servlet-api:2.5-20081211"
-libraries.jetty_util = "org.mortbay.jetty:jetty-util:6.1.26"
-libraries.jetty = "org.mortbay.jetty:jetty:6.1.26"
+libraries.servletApi = "javax.servlet:javax.servlet-api:3.1.0"
+libraries.jettyIo = "org.eclipse.jetty:jetty-io:${versions.jettyVersion}"
+libraries.jettySecurity = "org.eclipse.jetty:jetty-security:${versions.jettyVersion}"
+libraries.jettyServer = "org.eclipse.jetty:jetty-server:${versions.jettyVersion}"
+libraries.jettyServlet = "org.eclipse.jetty:jetty-servlet:${versions.jettyVersion}"
+libraries.jettyUtil = "org.eclipse.jetty:jetty-util:${versions.jettyVersion}"
+libraries.jetty = "org.eclipse.jetty:jetty-http:${versions.jettyVersion}"
+libraries.jettyWebApp = "org.eclipse.jetty:jetty-webapp:${versions.jettyVersion}"
 libraries.commons_codec = "commons-codec:commons-codec:1.11"
 libraries.jcifs = "org.samba.jcifs:jcifs:1.3.17"
 libraries.commons_httpclient = "org.apache.httpcomponents:httpclient:${versions.httpclient}"
diff --git a/subprojects/build-cache-http/build-cache-http.gradle b/subprojects/build-cache-http/build-cache-http.gradle
index 673fb31f0eaa1..4f9775a23e054 100644
--- a/subprojects/build-cache-http/build-cache-http.gradle
+++ b/subprojects/build-cache-http/build-cache-http.gradle
@@ -33,6 +33,7 @@ dependencies {
     compile project(':core')
     compile libraries.commons_httpclient
     compile libraries.commons_lang
+    testCompile(libraries.servletApi)
 
     integTestCompile libraries.jetty
 }
diff --git a/subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceTest.groovy b/subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceTest.groovy
index 22bccbc1948fc..c688e5d41b546 100644
--- a/subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceTest.groovy
+++ b/subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceTest.groovy
@@ -303,7 +303,8 @@ class HttpBuildCacheServiceTest extends Specification {
         server.expect("/cache/${key.hashCode}", false, [method], new HttpServer.ActionSupport("return ${httpCode} broken") {
             @Override
             void handle(HttpServletRequest request, HttpServletResponse response) {
-                response.sendError(httpCode, "broken")
+                //noinspection GrDeprecatedAPIUsage
+                response.setStatus(httpCode, "broken")
             }
         })
     }
diff --git a/subprojects/dependency-management/dependency-management.gradle b/subprojects/dependency-management/dependency-management.gradle
index 703b6c4af0280..0588e9d2e40b6 100644
--- a/subprojects/dependency-management/dependency-management.gradle
+++ b/subprojects/dependency-management/dependency-management.gradle
@@ -101,6 +98,7 @@ dependencies {
     integTestRuntime project(":testKit")
 
     testFixturesCompile project(":internalIntegTesting")
+    testFixturesCompile libraries.jettyWebApp
 }
 
 useTestFixtures()
diff --git a/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy b/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy
index e1bd3dcb67d1a..4bd11f03fe069 100644
--- a/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy
+++ b/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy
@@ -385,7 +385,7 @@ class DependencyUnresolvedModuleIntegrationTest extends AbstractHttpDependencyRe
         failure.assertHasCause("Failed to list versions for ${group}:${module}.")
         failure.assertHasCause("Could not get resource '${mavenHttpRepo.uri.toString()}/${group}/${module}/maven-metadata.xml'.")
         failure.assertHasCause("Unable to load Maven meta-data from ${mavenHttpRepo.uri.toString()}/${group}/${module}/maven-metadata.xml.")
-        failure.assertHasCause("Could not GET '${mavenHttpRepo.uri.toString()}/${group}/${module}/maven-metadata.xml'. Received status code 401 from server: unauthorized")
+        failure.assertHasCause("Could not GET '${mavenHttpRepo.uri.toString()}/${group}/${module}/maven-metadata.xml'. Received status code 401 from server: Unauthorized")
     }
 
     private void assertDependencyMetaDataReadTimeout(MavenModule module) {
@@ -404,7 +404,7 @@ class DependencyUnresolvedModuleIntegrationTest extends AbstractHttpDependencyRe
     private void assertDependencyMetaDataUnauthorizedError(MavenModule module) {
         failure.assertHasCause("Could not resolve ${mavenModuleCoordinates(module)}.")
         failure.assertHasCause("Could not get resource '${mavenHttpRepo.uri.toString()}/${mavenModuleRepositoryPath(module)}.pom'.")
-        failure.assertHasCause("Could not GET '${mavenHttpRepo.uri.toString()}/${mavenModuleRepositoryPath(module)}.pom'. Received status code 401 from server: unauthorized")
+        failure.assertHasCause("Could not GET '${mavenHttpRepo.uri.toString()}/${mavenModuleRepositoryPath(module)}.pom'. Received status code 401 from server: Unauthorized")
     }
 
     private void assertDependencyArtifactReadTimeout(MavenModule module) {
@@ -423,7 +423,7 @@ class DependencyUnresolvedModuleIntegrationTest extends AbstractHttpDependencyRe
     private void assertDependencyArtifactUnauthorizedError(MavenModule module) {
         failure.assertHasCause("Could not download ${module.artifactFile.name} (${mavenModuleCoordinates(module)})")
         failure.assertHasCause("Could not get resource '${mavenHttpRepo.uri.toString()}/${mavenModuleRepositoryPath(module)}.jar'.")
-        failure.assertHasCause("Could not GET '${mavenHttpRepo.uri.toString()}/${mavenModuleRepositoryPath(module)}.jar'. Received status code 401 from server: unauthorized")
+        failure.assertHasCause("Could not GET '${mavenHttpRepo.uri.toString()}/${mavenModuleRepositoryPath(module)}.jar'. Received status code 401 from server: Unauthorized")
     }
 
     private void assertDependencySkipped(MavenModule module) {
diff --git a/subprojects/internal-integ-testing/internal-integ-testing.gradle b/subprojects/internal-integ-testing/internal-integ-testing.gradle
index a4835ccd81e4c..a5400ae7cdaaa 100644
--- a/subprojects/internal-integ-testing/internal-integ-testing.gradle
+++ b/subprojects/internal-integ-testing/internal-integ-testing.gradle
@@ -21,12 +21,16 @@ plugins {
  * limitations under the License.
  */
 dependencies {
+    compile libraries.jettyWebApp
+
     compile libraries.groovy
     compile project(":internalTesting")
     compile project(':cli')
     compile project(':launcher')
     compile project(':native')
     compile libraries.jetty
+    compile libraries.jettySecurity
+
     compile module("org.littleshoot:littleproxy:1.1.0-beta1") {
         dependency libraries.slf4j_api
         dependency "io.netty:netty-all:4.0.23.Final"
diff --git a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/AuthScheme.groovy b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/AuthScheme.groovy
index 114e091b680d6..be16bc215ecdb 100644
--- a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/AuthScheme.groovy
+++ b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/AuthScheme.groovy
@@ -17,15 +17,19 @@
 
 package org.gradle.test.fixtures.server.http
 
-import org.mortbay.jetty.Response
-import org.mortbay.jetty.security.Authenticator
-import org.mortbay.jetty.security.BasicAuthenticator
-import org.mortbay.jetty.security.Constraint
-import org.mortbay.jetty.security.ConstraintMapping
-import org.mortbay.jetty.security.DigestAuthenticator
-import org.mortbay.jetty.security.SecurityHandler
-import org.mortbay.jetty.security.UserRealm
-
+import org.eclipse.jetty.http.HttpHeader
+import org.eclipse.jetty.security.Authenticator
+import org.eclipse.jetty.security.ConstraintMapping
+import org.eclipse.jetty.security.ConstraintSecurityHandler
+import org.eclipse.jetty.security.ServerAuthException
+import org.eclipse.jetty.security.authentication.BasicAuthenticator
+import org.eclipse.jetty.security.authentication.DigestAuthenticator
+import org.eclipse.jetty.server.Authentication
+import org.eclipse.jetty.server.ServletResponseHttpWrapper
+import org.eclipse.jetty.util.security.Constraint
+
+import javax.servlet.ServletRequest
+import javax.servlet.ServletResponse
 import javax.servlet.http.HttpServletResponse
 
 enum AuthScheme {
@@ -69,32 +73,39 @@ enum AuthScheme {
         protected Authenticator getAuthenticator() {
             return new BasicAuthenticator() {
                 @Override
-                void sendChallenge(UserRealm realm, Response response) throws IOException {
-                    response.sendError(HttpServletResponse.SC_NOT_FOUND)
+                Authentication validateRequest(ServletRequest req, ServletResponse res, boolean mandatory) throws ServerAuthException {
+                    def auth = super.validateRequest(req, new ServletResponseHttpWrapper(res), mandatory)
+                    if (!(auth instanceof Authentication.User)) {
+                        res.setHeader(HttpHeader.WWW_AUTHENTICATE.asString(), "basic realm=\"" + _loginService.getName() + '"')
+                        res.sendError(HttpServletResponse.SC_NOT_FOUND)
+                        return Authentication.SEND_CONTINUE
+                    }
+                    return super.validateRequest(req, res, mandatory)
                 }
             }
         }
     }
 
     abstract static class AuthSchemeHandler {
-        SecurityHandler createSecurityHandler(String path, TestUserRealm realm) {
+        ConstraintSecurityHandler createSecurityHandler(String path, TestUserRealm realm) {
             def constraintMapping = createConstraintMapping(path)
-            def securityHandler = new SecurityHandler()
-            securityHandler.userRealm = realm
-            securityHandler.constraintMappings = [constraintMapping] as ConstraintMapping[]
+            def securityHandler = new ConstraintSecurityHandler()
+            securityHandler.realmName = TestUserRealm.REALM_NAME
+            securityHandler.addConstraintMapping(constraintMapping)
             securityHandler.authenticator = authenticator
+            securityHandler.loginService = realm
             return securityHandler
         }
 
-        void addConstraint(SecurityHandler securityHandler, String path) {
-            securityHandler.constraintMappings = (securityHandler.constraintMappings as List) + createConstraintMapping(path)
+        void addConstraint(ConstraintSecurityHandler securityHandler, String path) {
+            securityHandler.addConstraintMapping(createConstraintMapping(path))
         }
 
         private ConstraintMapping createConstraintMapping(String path) {
             def constraint = new Constraint()
             constraint.name = constraintName()
             constraint.authenticate = true
-            constraint.roles = ['*'] as String[]
+            constraint.roles = [Constraint.ANY_ROLE, Constraint.ANY_AUTH, TestUserRealm.ROLE] as String[]
             def constraintMapping = new ConstraintMapping()
             constraintMapping.pathSpec = path
             constraintMapping.constraint = constraint
diff --git a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/DropConnectionFilter.groovy b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/DropConnectionFilter.groovy
index 01c4aa54bfc5b..f4705aea68a57 100644
--- a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/DropConnectionFilter.groovy
+++ b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/DropConnectionFilter.groovy
@@ -17,8 +17,8 @@
 package org.gradle.test.fixtures.server.http
 
 import com.google.common.io.ByteStreams
-import org.mortbay.util.IO
-import org.mortbay.util.URIUtil
+import org.eclipse.jetty.util.IO
+import org.eclipse.jetty.util.URIUtil
 
 import javax.servlet.Filter
 import javax.servlet.FilterChain
diff --git a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/HttpBuildCacheServer.groovy b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/HttpBuildCacheServer.groovy
index 27cbf40334389..b236adea6055f 100644
--- a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/HttpBuildCacheServer.groovy
+++ b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/HttpBuildCacheServer.groovy
@@ -17,13 +17,14 @@
 package org.gradle.test.fixtures.server.http
 
 import com.google.common.base.Preconditions
+import org.eclipse.jetty.server.Handler
+import org.eclipse.jetty.servlet.FilterHolder
+import org.eclipse.jetty.webapp.WebAppContext
 import org.gradle.test.fixtures.file.TestDirectoryProvider
 import org.gradle.test.fixtures.file.TestFile
 import org.junit.rules.ExternalResource
-import org.mortbay.jetty.Handler
-import org.mortbay.jetty.servlet.FilterHolder
-import org.mortbay.jetty.webapp.WebAppContext
-import org.mortbay.servlet.RestFilter
+
+import javax.servlet.DispatcherType
 
 class HttpBuildCacheServer extends ExternalResource implements HttpServerFixture {
     private final TestDirectoryProvider provider
@@ -34,6 +35,9 @@ class HttpBuildCacheServer extends ExternalResource implements HttpServerFixture
     HttpBuildCacheServer(TestDirectoryProvider provider) {
         this.provider = provider
         this.webapp = new WebAppContext()
+        // The following code is because of a problem under Windows: the file descriptors are kept open under JDK 11
+        // even after server shutdown, which prevents from deleting the test directory
+        this.webapp.setInitParameter("org.eclipse.jetty.servlet.Default.useFileMappedBuffer", "false");
     }
 
     TestFile getCacheDir() {
@@ -47,12 +51,13 @@ class HttpBuildCacheServer extends ExternalResource implements HttpServerFixture
 
     private void addFilters() {
         if (dropConnectionForPutBytes > -1) {
-            this.webapp.addFilter(new FilterHolder(new DropConnectionFilter(dropConnectionForPutBytes, this)), "/*", 1)
+            this.webapp.addFilter(new FilterHolder(new DropConnectionFilter(dropConnectionForPutBytes, this)), "/*", EnumSet.of(DispatcherType.REQUEST))
         }
         if (blockIncomingConnectionsForSeconds > 0) {
-            this.webapp.addFilter(new FilterHolder(new BlockFilter(blockIncomingConnectionsForSeconds)), "/*", 1)
+            this.webapp.addFilter(new FilterHolder(new BlockFilter(blockIncomingConnectionsForSeconds)), "/*", EnumSet.of(DispatcherType.REQUEST))
         }
-        this.webapp.addFilter(RestFilter, "/*", 1)
+        // TODO: Find Jetty 9 idiomatic way to get rid of this filter
+        this.webapp.addFilter(RestFilter, "/*", EnumSet.of(DispatcherType.REQUEST))
     }
 
     void dropConnectionForPutAfterBytes(long numBytes) {
diff --git a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/HttpServer.groovy b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/HttpServer.groovy
index 457c2306bcdf5..6fa9ee694fc79 100755
--- a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/HttpServer.groovy
+++ b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/HttpServer.groovy
@@ -20,7 +20,6 @@ import com.google.gson.Gson
 import com.google.gson.JsonElement
 import groovy.xml.MarkupBuilder
 import org.gradle.api.artifacts.repositories.PasswordCredentials
-import org.gradle.internal.BiAction
 import org.gradle.internal.hash.HashUtil
 import org.gradle.test.fixtures.server.ExpectOne
 import org.gradle.test.fixtures.server.ServerExpectation
@@ -29,16 +28,16 @@ import org.gradle.test.fixtures.server.ServerWithExpectations
 import org.gradle.test.matchers.UserAgentMatcher
 import org.gradle.util.GFileUtils
 import org.hamcrest.Matcher
-import org.mortbay.io.EndPoint
-import org.mortbay.jetty.Handler
-import org.mortbay.jetty.HttpHeaders
-import org.mortbay.jetty.HttpStatus
-import org.mortbay.jetty.MimeTypes
-import org.mortbay.jetty.Request
-import org.mortbay.jetty.handler.AbstractHandler
+import org.eclipse.jetty.server.Handler
+import org.eclipse.jetty.http.HttpHeader
+import org.eclipse.jetty.http.HttpStatus
+import org.eclipse.jetty.http.MimeTypes
+import org.eclipse.jetty.server.Request
+import org.eclipse.jetty.server.handler.AbstractHandler
 import org.slf4j.Logger
 import org.slf4j.LoggerFactory
 
+import javax.servlet.ServletException
 import javax.servlet.http.HttpServletRequest
 import javax.servlet.http.HttpServletResponse
 import java.util.concurrent.CountDownLatch
@@ -94,9 +93,10 @@ class HttpServer extends ServerWithExpectations implements HttpServerFixture {
     @Override
     Handler getCustomHandler() {
         return new AbstractHandler() {
-            void handle(String target, HttpServletRequest request, HttpServletResponse response, int dispatch) {
+            @Override
+            void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
                 String d = request.getQueryString()
-                if (request.handled) {
+                if (baseRequest.handled) {
                     return
                 }
                 onFailure(new AssertionError("Received unexpected ${request.method} request to ${target}."))
@@ -302,7 +302,7 @@ class HttpServer extends ServerWithExpectations implements HttpServerFixture {
     private Action broken() {
         new ActionSupport("return 500 broken") {
             void handle(HttpServletRequest request, HttpServletResponse response) {
-                response.sendError(500, "broken")
+                response.setStatus(500, "broken")
             }
         }
     }
@@ -387,7 +387,7 @@ class HttpServer extends ServerWithExpectations implements HttpServerFixture {
                 def file = srcFile
                 if (file.isFile()) {
                     response.setHeader("Content-Encoding", "gzip")
-                    response.setDateHeader(HttpHeaders.LAST_MODIFIED, srcFile.lastModified())
+                    response.setDateHeader(HttpHeader.LAST_MODIFIED.asString(), srcFile.lastModified())
                     def stream = new GZIPOutputStream(response.outputStream)
                     stream.write(file.bytes)
                     stream.close()
@@ -470,7 +470,7 @@ class HttpServer extends ServerWithExpectations implements HttpServerFixture {
 
     private sendFile(HttpServletResponse response, File file, Long lastModified, Long contentLength, String contentType) {
         if (sendLastModified) {
-            response.setDateHeader(HttpHeaders.LAST_MODIFIED, lastModified ?: file.lastModified())
+            response.setDateHeader(HttpHeader.LAST_MODIFIED.asString(), lastModified ?: file.lastModified())
         }
         def content = file.bytes
 
@@ -501,7 +501,7 @@ class HttpServer extends ServerWithExpectations implements HttpServerFixture {
             }
 
             if (value != null) {
-                response.addHeader(HttpHeaders.ETAG, value)
+                response.addHeader(HttpHeader.ETAG.asString(), value)
             }
         }
     }
@@ -528,7 +528,7 @@ class HttpServer extends ServerWithExpectations implements HttpServerFixture {
     /**
      * Expects one PUT request for the given URL. Writes the request content to the given file.
      */
-    void expectPut(String path, File destFile, int statusCode = HttpStatus.ORDINAL_200_OK, PasswordCredentials credentials = null, long expectedContentLength = -1) {
+    void expectPut(String path, File destFile, int statusCode = HttpStatus.OK_200, PasswordCredentials credentials = null, long expectedContentLength = -1) {
         def action = new ActionSupport("write request to $destFile.name (content length: $expectedContentLength) and return status $statusCode") {
             void handle(HttpServletRequest request, HttpServletResponse response) {
                 if (HttpServer.this.expectedUserAgent != null) {
@@ -654,7 +654,8 @@ class HttpServer extends ServerWithExpectations implements HttpServerFixture {
         HttpExpectOne expectation = new HttpExpectOne(action, methods, path)
         expectations << expectation
         add(path, matchPrefix, methods, new AbstractHandler() {
-            void handle(String target, HttpServletRequest request, HttpServletResponse response, int dispatch) {
+            @Override
+            void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
                 if (expectation.run) {
                     return
                 }
@@ -690,7 +691,8 @@ class HttpServer extends ServerWithExpectations implements HttpServerFixture {
 
     protected void allow(String path, boolean matchPrefix, Collection<String> methods, Action action) {
         add(path, matchPrefix, methods, new AbstractHandler() {
-            void handle(String target, HttpServletRequest request, HttpServletResponse response, int dispatch) {
+            @Override
+            void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
                 action.handle(request, response)
                 request.handled = true
             }
@@ -701,13 +703,14 @@ class HttpServer extends ServerWithExpectations implements HttpServerFixture {
         assert path.startsWith('/')
         def prefix = path == '/' ? '/' : path + '/'
         collection.addHandler(new AbstractHandler() {
-            void handle(String target, HttpServletRequest request, HttpServletResponse response, int dispatch) {
+            @Override
+            void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
                 if (methods != null && !methods.contains(request.method)) {
                     return
                 }
                 boolean match = request.pathInfo == path || (matchPrefix && request.pathInfo.startsWith(prefix))
                 if (match && !request.handled) {
-                    handler.handle(target, request, response, dispatch)
+                    handler.handle(target, baseRequest, request, response)
                 }
             }
         })
@@ -721,11 +724,8 @@ class HttpServer extends ServerWithExpectations implements HttpServerFixture {
      * Blocks on SSL handshake for 60 seconds.
      */
     void expectSslHandshakeBlocking() {
-        sslPreHandler = new BiAction<EndPoint, Request>() {
-            @Override
-            void execute(EndPoint endPoint, Request request) {
-                Thread.sleep(TimeUnit.SECONDS.toMillis(60))
-            }
+        sslPreHandler {
+            Thread.sleep(TimeUnit.SECONDS.toMillis(60))
         }
     }
 
diff --git a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/HttpServerFixture.groovy b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/HttpServerFixture.groovy
index c83f6efdca51a..279afdd30a49d 100644
--- a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/HttpServerFixture.groovy
+++ b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/HttpServerFixture.groovy
@@ -17,37 +17,34 @@
 package org.gradle.test.fixtures.server.http
 
 import com.google.common.collect.Sets
-import org.gradle.internal.BiAction
+import org.eclipse.jetty.http.HttpVersion
+import org.eclipse.jetty.server.Connector
+import org.eclipse.jetty.server.Handler
+import org.eclipse.jetty.server.HttpConfiguration
+import org.eclipse.jetty.server.HttpConnectionFactory
+import org.eclipse.jetty.server.Request
+import org.eclipse.jetty.server.SecureRequestCustomizer
+import org.eclipse.jetty.server.Server
+import org.eclipse.jetty.server.ServerConnector
+import org.eclipse.jetty.server.SslConnectionFactory
+import org.eclipse.jetty.server.handler.HandlerCollection
+import org.eclipse.jetty.util.ssl.SslContextFactory
 import org.gradle.util.ports.FixedAvailablePortAllocator
 import org.gradle.util.ports.PortAllocator
-import org.mortbay.io.EndPoint
-import org.mortbay.jetty.Connector
-import org.mortbay.jetty.Handler
-import org.mortbay.jetty.HttpHeaders
-import org.mortbay.jetty.Request
-import org.mortbay.jetty.Server
-import org.mortbay.jetty.bio.SocketConnector
-import org.mortbay.jetty.handler.AbstractHandler
-import org.mortbay.jetty.handler.HandlerCollection
-import org.mortbay.jetty.security.SecurityHandler
-import org.mortbay.jetty.security.SslSocketConnector
-
-import javax.servlet.http.HttpServletRequest
-import javax.servlet.http.HttpServletResponse
+
+import java.util.function.Consumer
 
 trait HttpServerFixture {
     private final PortAllocator portAllocator = FixedAvailablePortAllocator.instance
     private final Server server = new Server()
-    private Connector connector
-    private SslSocketConnector sslConnector
-    private final HandlerCollection collection = new HandlerCollection()
-    private TestUserRealm realm
-    private SecurityHandler securityHandler
-    private AuthScheme authenticationScheme = AuthScheme.BASIC
+    private ServerConnector connector
+    private ServerConnector sslConnector
+    private final SslPreHandler sslPreHandler = new SslPreHandler()
+    private final SecuredHandlerCollection securityHandlerWrapper = new SecuredHandlerCollection()
+
     private boolean logRequests = true
     private final Set<String> authenticationAttempts = Sets.newLinkedHashSet()
     private boolean configured
-    private int assignedPort
 
     Server getServer() {
         server
@@ -89,7 +86,7 @@ trait HttpServerFixture {
     abstract Handler getCustomHandler()
 
     HandlerCollection getCollection() {
-        return collection
+        return securityHandlerWrapper.handlers
     }
 
     Set<String> getAuthenticationAttempts() {
@@ -116,33 +113,29 @@ trait HttpServerFixture {
     }
 
     AuthScheme getAuthenticationScheme() {
-        return authenticationScheme
+        return securityHandlerWrapper.authenticationScheme
     }
 
     void setAuthenticationScheme(AuthScheme authenticationScheme) {
-        this.authenticationScheme = authenticationScheme
+        securityHandlerWrapper.authenticationScheme = authenticationScheme
     }
 
-    void setSslPreHandler(BiAction<EndPoint,Request> handler) {
-        server.connectors.each { connector ->
-            if (connector instanceof InterceptableSslSocketConnector) {
-                connector.sslPreHandler = handler
-            }
-        }
+    void sslPreHandler(Consumer<Request> consumer) {
+        sslPreHandler.registerCustomizer(consumer)
     }
 
     void start() {
         if (!configured) {
-            HandlerCollection handlers = new HandlerCollection()
+            HandlerCollection handlers = new HandlerCollection(true)
             handlers.addHandler(new LoggingHandler(authenticationAttempts, logRequests))
-            handlers.addHandler(collection)
+            handlers.addHandler(securityHandlerWrapper)
             handlers.addHandler(getCustomHandler())
             server.setHandler(handlers)
             configured = true
         }
 
-        assignedPort = portAllocator.assignPort()
-        connector = new SocketConnector()
+        def assignedPort = portAllocator.assignPort()
+        connector = new ServerConnector(server)
         connector.port = assignedPort
         server.addConnector(connector)
         server.start()
@@ -149,42 +142,22 @@ trait HttpServerFixture {
             // Has failed to start for some reason - try again
+            def port = connector.port
             server.removeConnector(connector)
             connector.stop()
-            portAllocator.releasePort(assignedPort)
-            assignedPort = portAllocator.assignPort()
-            connector = new SocketConnector()
-            connector.port = assignedPort
+            portAllocator.releasePort(port)
+            def newlyAssignedPort = portAllocator.assignPort()
+            connector = new ServerConnector(server)
+            connector.port = newlyAssignedPort
             server.addConnector(connector)
-            connector.start()
+            try {
+                connector.start()
+                return
+            } catch (e) {
+                println "Unable to start connector on port ${newlyAssignedPort}"
+            }
         }
         throw new AssertionError("SocketConnector failed to start.")
     }
 
-    private static class LoggingHandler extends AbstractHandler {
-        private final Set<String> authenticationAttempts
-        private final boolean logRequests
-
-        LoggingHandler(Set<String> authenticationAttempts, boolean logRequests) {
-            this.logRequests = logRequests
-            this.authenticationAttempts = authenticationAttempts
-        }
-
-        void handle(String target, HttpServletRequest request, HttpServletResponse response, int dispatch) {
-            String authorization = request.getHeader(HttpHeaders.AUTHORIZATION)
-            if (authorization != null) {
-                synchronized (authenticationAttempts) {
-                    authenticationAttempts << authorization.split(" ")[0]
-                }
-            } else {
-                synchronized (authenticationAttempts) {
-                    authenticationAttempts << "None"
-                }
-            }
-            if (logRequests) {
-                println("handling http request: $request.method $target")
-            }
-        }
-    }
-
     void stop() {
         if (sslConnector) {
             shutdownConnector(sslConnector)
@@ -218,41 +178,42 @@ trait HttpServerFixture {
         }
 
         server?.stop()
-        portAllocator.releasePort(assignedPort)
     }
 
     void reset() {
-        realm = null
-        collection.setHandlers()
+        securityHandlerWrapper.reset()
     }
 
     void enableSsl(String keyStore, String keyPassword, String trustStore = null, String trustPassword = null) {
-        sslConnector = new InterceptableSslSocketConnector()
-        sslConnector.keystore = keyStore
-        sslConnector.keyPassword = keyPassword
+        def sslContextFactory = new SslContextFactory.Server()
+        sslContextFactory.setIncludeProtocols("TLSv1.2")
+        sslContextFactory.setKeyStorePath(keyStore)
+        sslContextFactory.setKeyStorePassword(keyPassword)
         if (trustStore) {
-            sslConnector.needClientAuth = true
-            sslConnector.truststore = trustStore
-            sslConnector.trustPassword = trustPassword
+            sslContextFactory.needClientAuth = true
+            sslContextFactory.setTrustStorePath(trustStore)
         }
+        if (trustPassword) {
+            sslContextFactory.setTrustStorePassword(trustPassword)
+        }
+        def httpsConfig = new HttpConfiguration()
+        httpsConfig.addCustomizer(new SecureRequestCustomizer())
+        httpsConfig.addCustomizer(sslPreHandler)
+        def connectionFactory = new SslConnectionFactory(sslContextFactory, HttpVersion.HTTP_1_1.asString())
+        def httpConnectionFactory = new HttpConnectionFactory(httpsConfig)
+
+        sslConnector = new ServerConnector(server,
+            connectionFactory, httpConnectionFactory)
+        sslConnector.setPort(portAllocator.assignPort())
         server.addConnector(sslConnector)
         if (server.started) {
             sslConnector.start()
         }
+
     }
 
     void requireAuthentication(String path, String username, String password) {
-        if (realm != null) {
-            assert realm.username == username
-            assert realm.password == password
-            authenticationScheme.handler.addConstraint(securityHandler, path)
-        } else {
-            realm = new TestUserRealm()
-            realm.username = username
-            realm.password = password
-            securityHandler = authenticationScheme.handler.createSecurityHandler(path, realm)
-            collection.addHandler(securityHandler)
-        }
+        securityHandlerWrapper.requireAuthentication(path, username, password)
     }
 
     int getPort() {
@@ -263,20 +224,27 @@ trait HttpServerFixture {
         sslConnector.localPort
     }
 
-    private void shutdownConnector(Connector connector) {
+    private void shutdownConnector(ServerConnector connector) {
+        def port = connector.port
         connector.stop()
         connector.close()
         server?.removeConnector(connector)
+        portAllocator.releasePort(port)
     }
 }
 
-class InterceptableSslSocketConnector extends SslSocketConnector {
-    BiAction<EndPoint, Request> sslPreHandler
+class SslPreHandler implements HttpConfiguration.Customizer {
+
+    private final List<Consumer<Request>> consumers = []
+
+    void registerCustomizer(Consumer<Request> consumer) {
+        consumers << consumer
+    }
 
-    void customize(EndPoint endpoint, Request request) {
-        if (sslPreHandler) {
-            sslPreHandler.execute(endpoint, request)
+    @Override
+    void customize(Connector connector, HttpConfiguration channelConfig, Request request) {
+        consumers.each {
+            it.accept(request)
         }
-        super.customize(endpoint, request)
     }
 }
diff --git a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/LoggingHandler.groovy b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/LoggingHandler.groovy
new file mode 100644
index 0000000000000..b21b2f00b1d64
--- /dev/null
+++ b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/LoggingHandler.groovy
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.test.fixtures.server.http
+
+import groovy.transform.CompileStatic
+import org.eclipse.jetty.http.HttpHeader
+import org.eclipse.jetty.server.Request
+import org.eclipse.jetty.server.handler.AbstractHandler
+
+import javax.servlet.ServletException
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletResponse
+
+@CompileStatic
+class LoggingHandler extends AbstractHandler {
+    private final Set<String> authenticationAttempts
+    private final boolean logRequests
+
+    LoggingHandler(Set<String> authenticationAttempts, boolean logRequests) {
+        this.logRequests = logRequests
+        this.authenticationAttempts = authenticationAttempts
+    }
+
+    @Override
+    void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+        String authorization = getAuthorizationHeader(request)
+        if (authorization != null) {
+            synchronized (authenticationAttempts) {
+                authenticationAttempts << authorization.split(" ")[0]
+            }
+        } else {
+            synchronized (authenticationAttempts) {
+                authenticationAttempts << "None"
+            }
+        }
+        if (logRequests) {
+            println("handling http request: $request.method $target${request.queryString ? "?" + request.queryString : ''}")
+        }
+    }
+
+    protected static String getAuthorizationHeader(HttpServletRequest request) {
+        def header = request.getHeader(HttpHeader.AUTHORIZATION.asString())
+        return header
+    }
+}
diff --git a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/NtlmAuthenticator.groovy b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/NtlmAuthenticator.groovy
index 54c0c4f50ca6a..0c75e7452b8ed 100644
--- a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/NtlmAuthenticator.groovy
+++ b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/NtlmAuthenticator.groovy
@@ -16,64 +16,83 @@
 
 package org.gradle.test.fixtures.server.http
 
+import groovy.transform.CompileStatic
 import jcifs.http.NtlmSsp
 import jcifs.smb.NtlmPasswordAuthentication
-import org.mortbay.jetty.HttpHeaders
-import org.mortbay.jetty.Request
-import org.mortbay.jetty.Response
-import org.mortbay.jetty.security.Authenticator
-import org.mortbay.jetty.security.Credential
-import org.mortbay.jetty.security.UserRealm
-
+import org.eclipse.jetty.http.HttpHeader
+import org.eclipse.jetty.security.ServerAuthException
+import org.eclipse.jetty.security.UserAuthentication
+import org.eclipse.jetty.security.authentication.LoginAuthenticator
+import org.eclipse.jetty.server.Authentication
+import org.eclipse.jetty.server.HttpConnection
+import org.eclipse.jetty.util.security.Credential
+
+import javax.servlet.ServletRequest
+import javax.servlet.ServletResponse
+import javax.servlet.http.HttpServletRequest
 import javax.servlet.http.HttpServletResponse
-import java.security.Principal
 
-class NtlmAuthenticator implements Authenticator {
+@CompileStatic
+class NtlmAuthenticator extends LoginAuthenticator {
     static final String NTLM_AUTH_METHOD = 'NTLM'
 
+    // There is absolutely no doubt that no one should ever on a map like that
+    // and that a proper implementation of an NTLM authenticator shouldn't do this
+    // but those are test fixtures and I couldn't find a better way to do this
+    // without deeper understanding of the Jetty APIs. Feel free to provide an
+    // alternate solution
+    private final Map<HttpConnection, NtlmConnectionAuthentication> connections = [:]
+
+    @Override
+    String getAuthMethod() {
+        return NTLM_AUTH_METHOD
+    }
+
     @Override
-    Principal authenticate(UserRealm realm, String pathInContext, Request request, Response response) throws IOException {
-        NtlmConnectionAuthentication connectionAuth = request.connection.associatedObject
+    Authentication validateRequest(ServletRequest request, ServletResponse response, boolean mandatory) throws ServerAuthException {
+        HttpServletRequest httpRequest = (HttpServletRequest) request;
+        HttpServletResponse httpResponse = (HttpServletResponse) response
+        NtlmConnectionAuthentication connectionAuth = connections[HttpConnection.currentConnection]
 
         if (connectionAuth == null) {
             connectionAuth = new NtlmConnectionAuthentication(challenge: new byte[8])
             new Random().nextBytes(connectionAuth.challenge)
-
-            request.connection.associatedObject = connectionAuth
+            connections[HttpConnection.currentConnection] = connectionAuth
         }
 
-        if (connectionAuth.authenticated) {
-            request.authType = authMethod
-            request.userPrincipal = connectionAuth.principal
-
-            return connectionAuth.principal
+        if (connectionAuth.failed) {
+            httpResponse.setHeader(HttpHeader.WWW_AUTHENTICATE.asString(), "basic realm=\"" + _loginService.getName() + '"')
+            httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED)
+            return Authentication.SEND_CONTINUE
+        } else if (connectionAuth.user != null) {
+           return connectionAuth.user
         } else {
-            NtlmPasswordAuthentication authentication = NtlmSsp.authenticate(request, response, connectionAuth.challenge)
+            NtlmPasswordAuthentication authentication = NtlmSsp.authenticate(
+                (HttpServletRequest) request,
+                (HttpServletResponse) response,
+                connectionAuth.challenge)
 
             if (authentication != null) {
-                Principal principal = realm.authenticate(authentication.username, new TestNtlmCredentials(authentication, connectionAuth.challenge), request)
-
-                if (principal != null) {
-                    request.authType = authMethod
-                    request.userPrincipal = principal
-                    connectionAuth.principal = principal
+                def userIdentity = login(authentication.username, new TestNtlmCredentials(authentication, connectionAuth.challenge), request)
 
-                    return principal
+                if (userIdentity != null) {
+                    def user = new UserAuthentication(getAuthMethod(), userIdentity)
+                    connections[HttpConnection.currentConnection].user = user
+                    return user
                 } else {
-                    badCredentials(response)
+                    httpResponse.setHeader(HttpHeader.WWW_AUTHENTICATE.asString(), "basic realm=\"" + _loginService.getName() + '"')
+                    httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED)
+                    connections[HttpConnection.currentConnection].failed = true
+                    return Authentication.SEND_CONTINUE
                 }
             }
         }
+        return Authentication.SEND_CONTINUE
     }
 
     @Override
-    String getAuthMethod() {
-        return NTLM_AUTH_METHOD
-    }
-
-    private void badCredentials(Response response) {
-        response.setHeader(HttpHeaders.WWW_AUTHENTICATE, authMethod)
-        response.sendError(HttpServletResponse.SC_UNAUTHORIZED)
+    boolean secureResponse(ServletRequest request, ServletResponse response, boolean mandatory, Authentication.User validatedUser) throws ServerAuthException {
+        return true
     }
 
     private static class TestNtlmCredentials extends Credential {
@@ -91,7 +110,8 @@ class NtlmAuthenticator implements Authenticator {
                 byte[] hash = authentication.getAnsiHash(challenge)
                 byte[] clientChallenge = hash[16..-1]
 
-                return Arrays.equals(hash, NtlmPasswordAuthentication.getLMv2Response(authentication.domain, authentication.username, credentials, challenge, clientChallenge))
+                def response = NtlmPasswordAuthentication.getLMv2Response(authentication.domain, authentication.username, credentials, challenge, clientChallenge)
+                return Arrays.equals(hash, response)
             }
 
             return false
@@ -100,8 +120,10 @@ class NtlmAuthenticator implements Authenticator {
 
     private static class NtlmConnectionAuthentication {
         byte[] challenge
-        Principal principal
+        Authentication.User user
+        boolean failed
+
+        boolean isAuthenticated() { user != null}
 
-        boolean isAuthenticated() { principal != null}
     }
 }
diff --git a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/RestFilter.java b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/RestFilter.java
new file mode 100644
index 0000000000000..ee342c4c44c9b
--- /dev/null
+++ b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/RestFilter.java
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.test.fixtures.server.http;
+
+import org.eclipse.jetty.util.IO;
+import org.eclipse.jetty.util.URIUtil;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.UnavailableException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+
+/**
+ * <p>
+ * Support for HTTP PUT and DELETE methods.
+ * </p>
+ * <p><b>
+ * THIS FILTER SHOULD ONLY BE USED WITH VERY GOOD SECURITY CONSTRAINTS!
+ * </b></p>
+ *
+ * <p>
+ * If the filter init parameter maxPutSize is set to a positive integer, then
+ * only puts of known size less than maxPutSize will be accepted.
+ * </p>
+ *
+ */
+// Copied from org.mortbay.servlet.RestFilter
+public class RestFilter implements Filter {
+    private static final String HTTP_METHOD_PUT = "PUT";
+    private static final String HTTP_METHOD_GET = "GET";
+    private static final String HTTP_METHOD_DELETE = "DELETE";
+
+    private FilterConfig filterConfig;
+    private long _maxPutSize;
+
+    /* ------------------------------------------------------------ */
+    /* (non-Javadoc)
+     * @see javax.servlet.Filter#init(javax.servlet.FilterConfig)
+     */
+    public void init(FilterConfig filterConfig) throws UnavailableException {
+        this.filterConfig = filterConfig;
+        String tmp = filterConfig.getInitParameter("maxPutSize");
+        if (tmp != null) {
+            _maxPutSize = Long.parseLong(tmp);
+        }
+    }
+
+    /* ------------------------------------------------------------ */
+
+    /**
+     *
+     */
+    private File locateFile(HttpServletRequest request) {
+        return new File(filterConfig.getServletContext().getRealPath(URIUtil.addPaths(request.getServletPath(), request.getPathInfo())));
+    }
+
+    /* ------------------------------------------------------------ */
+    /* (non-Javadoc)
+     * @see javax.servlet.Filter#doFilter(javax.servlet.ServletRequest, javax.servlet.ServletResponse, javax.servlet.FilterChain)
+     */
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        if (!(request instanceof HttpServletRequest && response instanceof HttpServletResponse)) {
+            chain.doFilter(request, response);
+            return;
+        }
+
+        HttpServletRequest httpRequest = (HttpServletRequest) request;
+        HttpServletResponse httpResponse = (HttpServletResponse) response;
+
+        if (httpRequest.getMethod().equals(HTTP_METHOD_GET)) {
+            chain.doFilter(httpRequest, httpResponse);
+        } else if (httpRequest.getMethod().equals(HTTP_METHOD_PUT)) {
+            doPut(httpRequest, httpResponse);
+        } else if (httpRequest.getMethod().equals(HTTP_METHOD_DELETE)) {
+            doDelete(httpRequest, httpResponse);
+        } else {
+            chain.doFilter(httpRequest, httpResponse);
+        }
+    }
+    /* ------------------------------------------------------------ */
+
+    /**
+     *
+     */
+    protected void doPut(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        File file = locateFile(request);
+
+        if (file.exists()) {
+            boolean success = file.delete(); // replace file if it exists
+            if (!success) {
+                response.sendError(HttpServletResponse.SC_FORBIDDEN);
+                return;
+            }
+        }
+
+        FileOutputStream out = new FileOutputStream(file);
+        try {
+            if (_maxPutSize > 0) {
+                int length = request.getContentLength();
+                if (length > _maxPutSize) {
+                    response.sendError(HttpServletResponse.SC_FORBIDDEN);
+                    return;
+                }
+                IO.copy(request.getInputStream(), out, _maxPutSize);
+            } else {
+                IO.copy(request.getInputStream(), out);
+            }
+        } finally {
+            out.close();
+        }
+
+        response.setStatus(HttpServletResponse.SC_NO_CONTENT);
+    }
+
+    /* ------------------------------------------------------------ */
+
+    /**
+     *
+     */
+    protected void doDelete(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        File file = locateFile(request);
+
+        if (!file.exists()) {
+            response.sendError(HttpServletResponse.SC_NOT_FOUND);
+            return;
+        }
+
+        boolean success = IO.delete(file); // actual delete operation
+
+        if (success) {
+            response.setStatus(HttpServletResponse.SC_NO_CONTENT);
+        } else {
+            response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    public void destroy() {
+        // nothing to destroy
+    }
+}
diff --git a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/SecuredHandlerCollection.groovy b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/SecuredHandlerCollection.groovy
new file mode 100644
index 0000000000000..9cdf2a460c19d
--- /dev/null
+++ b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/SecuredHandlerCollection.groovy
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.test.fixtures.server.http
+
+import groovy.transform.CompileStatic
+import org.eclipse.jetty.security.ConstraintSecurityHandler
+import org.eclipse.jetty.server.Handler
+import org.eclipse.jetty.server.Request
+import org.eclipse.jetty.server.Server
+import org.eclipse.jetty.server.handler.HandlerCollection
+
+import javax.servlet.ServletException
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletResponse
+
+@CompileStatic
+class SecuredHandlerCollection implements Handler {
+    private final HandlerCollection handlers = new HandlerCollection(true)
+    private ConstraintSecurityHandler securityHandler
+    private TestUserRealm realm
+
+    AuthScheme authenticationScheme = AuthScheme.BASIC
+
+    private Handler delegate = handlers
+
+    SecuredHandlerCollection() {
+    }
+
+    HandlerCollection getHandlers() {
+        handlers
+    }
+
+    void reset() {
+        realm = null
+        securityHandler?.stop()
+        securityHandler = null
+        handlers.stop()
+        delegate = handlers
+        handlers.setHandlers(new Handler[0])
+        handlers.start()
+    }
+
+    @Override
+    void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+        delegate.handle(target, baseRequest, request, response)
+    }
+
+    @Override
+    void setServer(Server server) {
+        delegate.server = server
+    }
+
+    @Override
+    Server getServer() {
+        delegate.server
+    }
+
+    @Override
+    void destroy() {
+        delegate.destroy()
+    }
+
+    void requireAuthentication(String path, String username, String password) {
+        if (realm != null) {
+            assert realm.username == username
+            assert realm.password == password
+            authenticationScheme.handler.addConstraint(securityHandler, path)
+        } else {
+            realm = new TestUserRealm(username, password)
+            securityHandler = authenticationScheme.handler.createSecurityHandler(path, realm)
+            def server = handlers.server
+            handlers.stop()
+            securityHandler.handler = handlers
+            delegate = securityHandler
+            delegate.server = server
+            delegate.start()
+        }
+    }
+
+    @Override
+    void start() throws Exception {
+        delegate.start()
+    }
+
+    @Override
+    void stop() throws Exception {
+        delegate.stop()
+    }
+
+    @Override
+    boolean isRunning() {
+        delegate.running
+    }
+
+    @Override
+    boolean isStarted() {
+        delegate.started
+    }
+
+    @Override
+    boolean isStarting() {
+        delegate.starting
+    }
+
+    @Override
+    boolean isStopping() {
+        delegate.stopping
+    }
+
+    @Override
+    boolean isStopped() {
+        delegate.stopped
+    }
+
+    @Override
+    boolean isFailed() {
+        delegate.failed
+    }
+
+    @Override
+    void addLifeCycleListener(Listener listener) {
+        delegate.addLifeCycleListener(listener)
+    }
+
+    @Override
+    void removeLifeCycleListener(Listener listener) {
+        delegate.removeLifeCycleListener(listener)
+    }
+}
diff --git a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/TestUserRealm.groovy b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/TestUserRealm.groovy
index c3f4a7fc16820..29a10af25b3eb 100644
--- a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/TestUserRealm.groovy
+++ b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/TestUserRealm.groovy
@@ -14,57 +14,29 @@
  * limitations under the License.
  */
 
-package org.gradle.test.fixtures.server.http;
+package org.gradle.test.fixtures.server.http
 
-import org.mortbay.jetty.Request;
-import org.mortbay.jetty.security.Password;
-import org.mortbay.jetty.security.UserRealm;
+import groovy.transform.CompileStatic
+import org.eclipse.jetty.security.HashLoginService
+import org.eclipse.jetty.security.UserStore
+import org.eclipse.jetty.util.security.Credential
 
-import java.security.Principal;
+@CompileStatic
+class TestUserRealm extends HashLoginService {
+    public static final String REALM_NAME = 'test'
+    public static final String ROLE = 'test'
+    private final UserStore userStore
 
-class TestUserRealm implements UserRealm {
-    String username
-    String password
+    final String username
+    final String password
 
-    Principal authenticate(String username, Object credentials, Request request) {
-        Password passwordCred = new Password(password)
-        if (username == this.username && passwordCred.check(credentials)) {
-            return getPrincipal(username)
-        }
-        return null
+    TestUserRealm(String username, String password) {
+        super(REALM_NAME)
+        userStore = new UserStore()
+        setUserStore(userStore)
+        this.username = username
+        this.password = password
+        userStore.addUser(username, Credential.getCredential(password), [ROLE] as String[])
     }
 
-    String getName() {
-        return "test"
-    }
-
-    Principal getPrincipal(String username) {
-        return new Principal() {
-            String getName() {
-                return username
-            }
-        }
-    }
-
-    boolean reauthenticate(Principal user) {
-        return false
-    }
-
-    boolean isUserInRole(Principal user, String role) {
-        return false
-    }
-
-    void disassociate(Principal user) {
-    }
-
-    Principal pushRole(Principal user, String role) {
-        return user
-    }
-
-    Principal popRole(Principal user) {
-        return user
-    }
-
-    void logout(Principal user) {
-    }
 }
diff --git a/subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/WithExternalRepository.groovy b/subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/WithExternalRepository.groovy
index 0af099d58ac5a..96470ae14ac83 100644
--- a/subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/WithExternalRepository.groovy
+++ b/subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/WithExternalRepository.groovy
@@ -19,11 +19,10 @@ package org.gradle.performance
 import groovy.transform.CompileStatic
 import groovy.transform.SelfType
 import org.apache.mina.util.AvailablePortFinder
+import org.eclipse.jetty.server.Server
+import org.eclipse.jetty.util.resource.Resource
+import org.eclipse.jetty.webapp.WebAppContext
 import org.gradle.performance.fixture.TestProjectLocator
-import org.mortbay.jetty.Server
-import org.mortbay.jetty.servlet.Context
-import org.mortbay.jetty.webapp.WebAppContext
-import org.mortbay.resource.Resource
 
 @CompileStatic
 @SelfType(AbstractCrossVersionPerformanceTest)
@@ -36,7 +35,7 @@ trait WithExternalRepository {
 
     abstract AbstractCrossVersionPerformanceTestRunner getRunner()
 
-    Context createContext() {
+    WebAppContext createContext() {
         def context = new WebAppContext()
         context.setContextPath("/")
         context.setBaseResource(Resource.newResource(repoDir.getAbsolutePath()))
@@ -47,10 +46,10 @@ trait WithExternalRepository {
         try {
             serverPort = AvailablePortFinder.getNextAvailable(5000)
             server = new Server(serverPort)
-            Context context = createContext()
+            WebAppContext context = createContext()
             context.setContextPath("/")
             context.setBaseResource(Resource.newResource(repoDir.getAbsolutePath()))
-            server.addHandler(context)
+            server.insertHandler(context)
             server.start()
         } catch (IllegalArgumentException ex) {
             server = null // repository not found, probably running on coordinator. If not, error will be caught later
diff --git a/subprojects/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishHttpIntegTest.groovy b/subprojects/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishHttpIntegTest.groovy
index 4df5072de9a37..768d39fca6c20 100644
--- a/subprojects/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishHttpIntegTest.groovy
+++ b/subprojects/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishHttpIntegTest.groovy
@@ -16,6 +16,7 @@
 
 package org.gradle.api.publish.ivy
 
+import org.eclipse.jetty.http.HttpStatus
 import org.gradle.integtests.fixtures.executer.ProgressLoggingFixture
 import org.gradle.internal.jvm.Jvm
 import org.gradle.test.fixtures.file.TestFile
@@ -29,7 +30,6 @@ import org.gradle.test.fixtures.server.http.IvyHttpRepository
 import org.gradle.util.Requires
 import org.hamcrest.Matchers
 import org.junit.Rule
-import org.mortbay.jetty.HttpStatus
 import spock.lang.Issue
 import spock.lang.Unroll
 
@@ -93,8 +93,8 @@
         and:
         module.jar.expectPut()
         module.jar.sha1.expectPut()
-        module.ivy.expectPut(HttpStatus.ORDINAL_201_Created)
-        module.ivy.sha1.expectPut(HttpStatus.ORDINAL_201_Created)
+        module.ivy.expectPut(HttpStatus.CREATED_201)
+        module.ivy.sha1.expectPut(HttpStatus.CREATED_201)
         module.moduleMetadata.expectPut()
         module.moduleMetadata.sha1.expectPut()
 
diff --git a/subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/corefeature/ParallelDownloadsPerformanceTest.groovy b/subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/corefeature/ParallelDownloadsPerformanceTest.groovy
index e4e48bc89f58e..671ed5e30b9b1 100644
--- a/subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/corefeature/ParallelDownloadsPerformanceTest.groovy
+++ b/subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/corefeature/ParallelDownloadsPerformanceTest.groovy
@@ -16,16 +16,15 @@
 
 package org.gradle.performance.regression.corefeature
 
+import org.eclipse.jetty.webapp.WebAppContext
 import org.gradle.performance.AbstractCrossVersionPerformanceTest
 import org.gradle.performance.WithExternalRepository
 import org.gradle.performance.fixture.BuildExperimentInvocationInfo
 import org.gradle.performance.fixture.BuildExperimentListener
 import org.gradle.performance.fixture.BuildExperimentListenerAdapter
 import org.gradle.performance.measure.MeasuredOperation
-import org.mortbay.jetty.Handler
-import org.mortbay.jetty.servlet.Context
-import org.mortbay.jetty.webapp.WebAppContext
 
+import javax.servlet.DispatcherType
 import javax.servlet.Filter
 import javax.servlet.FilterChain
 import javax.servlet.FilterConfig
@@ -105,9 +104,9 @@ class ParallelDownloadsPerformanceTest extends AbstractCrossVersionGradleInterna
 
 
     @Override
-    Context createContext() {
+    WebAppContext createContext() {
         def context = new WebAppContext()
-        context.addFilter(SimulatedDownloadLatencyFilter, '/*', Handler.DEFAULT)
+        context.addFilter(SimulatedDownloadLatencyFilter, '/*', EnumSet.of(DispatcherType.REQUEST))
         context
     }
 
diff --git a/subprojects/resources-gcs/src/integTest/groovy/org/gradle/integtests/resource/gcs/fixtures/GcsServer.groovy b/subprojects/resources-gcs/src/integTest/groovy/org/gradle/integtests/resource/gcs/fixtures/GcsServer.groovy
index 8b7ec11fb20c6..d11db0800cb6c 100644
--- a/subprojects/resources-gcs/src/integTest/groovy/org/gradle/integtests/resource/gcs/fixtures/GcsServer.groovy
+++ b/subprojects/resources-gcs/src/integTest/groovy/org/gradle/integtests/resource/gcs/fixtures/GcsServer.groovy
@@ -16,6 +16,8 @@
 
 package org.gradle.integtests.resource.gcs.fixtures
 
+import org.eclipse.jetty.server.Request
+import org.eclipse.jetty.server.handler.AbstractHandler
 import org.gradle.integtests.resource.gcs.fixtures.stub.HttpStub
 import org.gradle.integtests.resource.gcs.fixtures.stub.StubRequest
 import org.gradle.test.fixtures.file.TestDirectoryProvider
@@ -25,9 +27,8 @@ import org.joda.time.DateTimeZone
 import org.joda.time.format.DateTimeFormat
 import org.joda.time.format.DateTimeFormatter
 import org.joda.time.tz.FixedDateTimeZone
-import org.mortbay.jetty.Request
-import org.mortbay.jetty.handler.AbstractHandler
 
+import javax.servlet.ServletException
 import javax.servlet.http.HttpServletRequest
 import javax.servlet.http.HttpServletResponse
 import java.security.MessageDigest
@@ -177,7 +178,7 @@ class GcsServer extends HttpServer implements RepositoryServer {
                 ]
                 body = {
                     """
-                    { 
+                    {
                         "etag": "${calculateEtag(file)}",
                         "size": "0",
                         "updated": "${RCF_3339_DATE_FORMAT.print(file.lastModified())}",
@@ -427,7 +428,7 @@ class GcsServer extends HttpServer implements RepositoryServer {
         add(httpStub, stubAction(httpStub))
     }
 
-    private HttpServer.ActionSupport stubAction(HttpStub httpStub) {
+    private static HttpServer.ActionSupport stubAction(HttpStub httpStub) {
         new HttpServer.ActionSupport("Generic stub handler") {
             void handle(HttpServletRequest request, HttpServletResponse response) {
                 if (httpStub.request.body) {
@@ -448,20 +449,23 @@ class GcsServer extends HttpServer implements RepositoryServer {
         HttpServer.HttpExpectOne expectation = new HttpServer.HttpExpectOne(action, [httpStub.request.method], httpStub.request.path)
         expectations << expectation
         addHandler(new AbstractHandler() {
-            void handle(String target, HttpServletRequest request, HttpServletResponse response, int dispatch) {
+            @Override
+            void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+
                 if (requestMatches(httpStub, request)) {
                     assertRequest(httpStub, request)
                     if (expectation.run) {
                         println("This expectation for the request [${request.method} :${request.pathInfo}] was already handled - skipping")
                         return
                     }
-                    if (!((Request) request).isHandled()) {
+                    if (!baseRequest.isHandled()) {
                         expectation.run = true
                         action.handle(request, response)
-                        ((Request) request).setHandled(true)
+                        baseRequest.setHandled(true)
                     }
                 }
             }
+
         })
     }
 
diff --git a/subprojects/resources-gcs/src/integTest/groovy/org/gradle/integtests/resource/gcs/maven/MavenPublishGcsIntegrationTest.groovy b/subprojects/resources-gcs/src/integTest/groovy/org/gradle/integtests/resource/gcs/maven/MavenPublishGcsIntegrationTest.groovy
index a79dc5981e448..ceb0fb2fc2da0 100644
--- a/subprojects/resources-gcs/src/integTest/groovy/org/gradle/integtests/resource/gcs/maven/MavenPublishGcsIntegrationTest.groovy
+++ b/subprojects/resources-gcs/src/integTest/groovy/org/gradle/integtests/resource/gcs/maven/MavenPublishGcsIntegrationTest.groovy
@@ -21,10 +21,13 @@ import org.gradle.integtests.fixtures.publish.maven.AbstractMavenPublishIntegTes
 import org.gradle.integtests.resource.gcs.fixtures.GcsArtifact
 import org.gradle.integtests.resource.gcs.fixtures.GcsServer
 import org.gradle.integtests.resource.gcs.fixtures.MavenGcsRepository
+import org.gradle.util.Requires
+import org.gradle.util.TestPrecondition
 import org.junit.Rule
 
 import static org.gradle.internal.resource.transport.gcp.gcs.GcsConnectionProperties.*
 
+@Requires(TestPrecondition.NOT_WINDOWS)
 class MavenPublishGcsIntegrationTest extends AbstractMavenPublishIntegTest {
     @Rule
     public GcsServer server = new GcsServer(temporaryFolder)
diff --git a/subprojects/resources-http/resources-http.gradle b/subprojects/resources-http/resources-http.gradle
index 86f62780558ef..c8bd2e415ee33 100644
--- a/subprojects/resources-http/resources-http.gradle
+++ b/subprojects/resources-http/resources-http.gradle
@@ -36,7 +36,7 @@ dependencies {
     implementation libraries.xerces
     implementation libraries.nekohtml
 
-    testImplementation libraries.jetty
+    testImplementation libraries.jettyWebApp
 }
 
 useTestFixtures()
diff --git a/subprojects/resources-http/src/integTest/groovy/org/gradle/internal/resource/transport/http/CookieHeaderTest.groovy b/subprojects/resources-http/src/integTest/groovy/org/gradle/internal/resource/transport/http/CookieHeaderTest.groovy
index 2043409714995..5542fa699d7eb 100644
--- a/subprojects/resources-http/src/integTest/groovy/org/gradle/internal/resource/transport/http/CookieHeaderTest.groovy
+++ b/subprojects/resources-http/src/integTest/groovy/org/gradle/internal/resource/transport/http/CookieHeaderTest.groovy
@@ -24,7 +24,6 @@ import org.gradle.api.logging.LogLevel
 import org.junit.Rule
 import spock.lang.Unroll
 
-import javax.servlet.http.Cookie
 import javax.servlet.http.HttpServletRequest
 import javax.servlet.http.HttpServletResponse
 
@@ -47,7 +46,7 @@ class CookieHeaderTest extends Specification {
     def "cookie header with attributes #attributes can be parsed"() {
         httpServer.start()
         httpServer.expect("/cookie", ['GET'],
-            new RespondWithCookieAction("some; ${attributes}"))
+            new RespondWithCookieAction("some", attributes))
         when:
         client.performGet("${httpServer.address}/cookie", false)
 
@@ -67,15 +66,17 @@ class CookieHeaderTest extends Specification {
 
 class RespondWithCookieAction extends HttpServer.ActionSupport {
     private final String cookie
+    private final String attributes
 
-    RespondWithCookieAction(String cookie) {
+    RespondWithCookieAction(String cookie, String attributes) {
         super("Return cookie header ${cookie}")
         this.cookie = cookie
+        this.attributes = attributes
     }
 
     @Override
     void handle(HttpServletRequest request, HttpServletResponse response) {
-        response.addCookie(new Cookie("cookie_name", cookie))
+        response.addHeader("Set-Cookie", "cookie_name=$cookie; $attributes")
         String message = "Cookie sent"
         response.setContentLength(message.bytes.length)
         response.setContentType("text/html")
diff --git a/subprojects/resources-s3/src/integTest/groovy/org/gradle/integtests/resource/s3/fixtures/S3Server.groovy b/subprojects/resources-s3/src/integTest/groovy/org/gradle/integtests/resource/s3/fixtures/S3Server.groovy
index 33e4b86cb6b17..c6332aed945e4 100644
--- a/subprojects/resources-s3/src/integTest/groovy/org/gradle/integtests/resource/s3/fixtures/S3Server.groovy
+++ b/subprojects/resources-s3/src/integTest/groovy/org/gradle/integtests/resource/s3/fixtures/S3Server.groovy
@@ -17,6 +17,8 @@
 package org.gradle.integtests.resource.s3.fixtures
 
 import groovy.xml.StreamingMarkupBuilder
+import org.eclipse.jetty.server.Request
+import org.eclipse.jetty.server.handler.AbstractHandler
 import org.gradle.integtests.resource.s3.fixtures.stub.HttpStub
 import org.gradle.integtests.resource.s3.fixtures.stub.StubRequest
 import org.gradle.test.fixtures.file.TestDirectoryProvider
@@ -27,9 +29,8 @@ import org.joda.time.DateTimeZone
 import org.joda.time.format.DateTimeFormat
 import org.joda.time.format.DateTimeFormatter
 import org.joda.time.tz.FixedDateTimeZone
-import org.mortbay.jetty.Request
-import org.mortbay.jetty.handler.AbstractHandler
 
+import javax.servlet.ServletException
 import javax.servlet.http.HttpServletRequest
 import javax.servlet.http.HttpServletResponse
 import java.security.MessageDigest
@@ -472,7 +473,9 @@ class S3Server extends HttpServer implements RepositoryServer {
         HttpServer.HttpExpectOne expectation = new HttpServer.HttpExpectOne(action, [httpStub.request.method], httpStub.request.path)
         expectations << expectation
         addHandler(new AbstractHandler() {
-            void handle(String target, HttpServletRequest request, HttpServletResponse response, int dispatch) {
+            @Override
+            void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+
                 if (requestMatches(httpStub, request)) {
                     assertRequest(httpStub, request)
                     if (expectation.run) {
diff --git a/subprojects/soak/src/integTest/groovy/org/gradle/resolve/DependencyResolutionStressTest.groovy b/subprojects/soak/src/integTest/groovy/org/gradle/resolve/DependencyResolutionStressTest.groovy
index c632d9a58ad7a..a5b875fdd191c 100644
--- a/subprojects/soak/src/integTest/groovy/org/gradle/resolve/DependencyResolutionStressTest.groovy
+++ b/subprojects/soak/src/integTest/groovy/org/gradle/resolve/DependencyResolutionStressTest.groovy
@@ -16,6 +16,12 @@
 
 package org.gradle.resolve
 
+import org.eclipse.jetty.http.HttpHeader
+import org.eclipse.jetty.server.Request
+import org.eclipse.jetty.server.Server
+import org.eclipse.jetty.server.ServerConnector
+import org.eclipse.jetty.server.handler.AbstractHandler
+import org.eclipse.jetty.server.handler.HandlerCollection
 import org.gradle.integtests.fixtures.executer.GradleDistribution
 import org.gradle.integtests.fixtures.executer.GradleExecuter
 import org.gradle.integtests.fixtures.executer.IntegrationTestBuildContext
@@ -25,13 +31,9 @@ import org.gradle.test.fixtures.ConcurrentTestUtil
 import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
 import org.junit.Rule
 import org.junit.rules.ExternalResource
-import org.mortbay.jetty.Connector
-import org.mortbay.jetty.HttpHeaders
-import org.mortbay.jetty.Server
-import org.mortbay.jetty.handler.AbstractHandler
-import org.mortbay.jetty.nio.SelectChannelConnector
 import spock.lang.Specification
 
+import javax.servlet.ServletException
 import javax.servlet.http.HttpServletRequest
 import javax.servlet.http.HttpServletResponse
 import java.util.zip.ZipEntry
@@ -106,14 +108,15 @@ task check {
         private static final String METADATA_FILE_PATH = '/org.gradle/changing/1.0/ivy-1.0.xml'
         private static final String JAR_FILE_PATH = '/org.gradle/changing/1.0/changing-1.0.jar'
         private final Server server = new Server(0)
-        private final SelectChannelConnector connector = new SelectChannelConnector()
+        private final ServerConnector connector = new ServerConnector(server)
         private final Resources resources = new Resources()
 
         @Override
         protected void before() {
-            server.setConnectors([connector] as Connector[])
-            server.addHandler(new AbstractHandler() {
-                void handle(String target, HttpServletRequest request, HttpServletResponse response, int dispatch) {
+            server.addConnector(connector)
+            def handler = new AbstractHandler() {
+                @Override
+                void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
                     println "* Handling $request.method $request.pathInfo"
                     if (request.method == GET_METHOD && request.pathInfo == METADATA_FILE_PATH) {
                         handleGetIvy(response)
@@ -129,7 +132,8 @@ task check {
                         request.handled = true
                     }
                 }
-            })
+            }
+            server.setHandler(new HandlerCollection(false, handler))
             server.start()
         }
 
@@ -162,8 +166,8 @@ task check {
             provideHeadersForResource(response, resources.jar)
         }
 
-        private void provideHeadersForResource(HttpServletResponse response, Resource resource) {
-            response.setDateHeader(HttpHeaders.LAST_MODIFIED, resource.lastModified)
+        private static void provideHeadersForResource(HttpServletResponse response, Resource resource) {
+            response.setDateHeader(HttpHeader.LAST_MODIFIED.asString(), resource.lastModified)
             response.setContentLength(resource.contentLength)
             response.setContentType(resource.contentType)
         }
